sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -293,20 +294,118 @@ case class Substring(str: Expression, pos: Expression, len: Expression)
 }
 
 /**
- * A function that return the length of the given string expression.
+ * A function that return the length of the given string or binary expression.
  */
-case class StringLength(child: Expression) extends UnaryExpression with ExpectsInputTypes {
+case class Length(child: Expression) extends UnaryExpression with ExpressionConstraint {
   override def dataType: DataType = IntegerType
-  override def expectedChildTypes: Seq[DataType] = Seq(StringType)
+  def constraint: Seq[DataTypeConstraint] =
+    DataTypeConstraint(AcceptSpecifiedType(Set(StringType, BinaryType))) :: Nil
+
+  @transient
+  private lazy val function: Any => Any = child.dataType match {
+    case StringType => (s: Any) => s.asInstanceOf[UTF8String].length()
+    case BinaryType => (s: Any) => s.asInstanceOf[Array[Byte]].length
+    case NullType => (s: Any) => null
```

Comment:
Since you write `if (value == null) null else function(value)`, do we need to handle `NullType`?